sql/analyzer/validation_rules.go

```diff
@@ -232,31 +232,44 @@ func validateGroupBy(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, se
 		return n, transform.SameTree, nil
 	}
 
-	switch n := n.(type) {
-	case *plan.GroupBy:
-		// Allow the parser use the GroupBy node to eval the aggregation functions
-		// for sql statements that don't make use of the GROUP BY expression.
-		if len(n.GroupByExprs) == 0 {
-			return n, transform.SameTree, nil
-		}
+	var err error
+	var parent sql.Node
+	transform.Inspect(n, func(n sql.Node) bool {
+		if gb, ok := n.(*plan.GroupBy); ok {
+			if _, ok := parent.(*plan.Having); ok {
+				return true
+			}
+			if _, ok := parent.(*plan.Project); ok {
+				return true
+			}
+			if _, ok := parent.(*plan.Sort); ok {
+				return true
```

Comment:
Initially, this rule only ran when the top-most node was a `GroupBy`, which led to a lot of queries not erroring when they should've. Then, I changed it to use a `transform.Inspect`, which cause a bunch of queries to error. I think some of the errors are caused by aliases confusing the rule, and others might just be wrong...

Reply:
I think some queries are broken because we don't do this: https://dev.mysql.com/doc/refman/8.0/en/group-by-handling.html#:~:text=The%20query%20is%20valid%20if%20name%20is%20a%20primary%20key